wagtail/contrib/redirects/views.py

```diff
@@ -36,66 +37,88 @@
+    model = Redirect
+    header_icon = "redirects"
+    add_item_label = gettext_lazy("Add redirect")
+    context_object_name = "redirects"
+    index_url_name = "wagtailredirects:index"
+    add_url_name = "wagtailredirects:add"
+    edit_url_name = "wagtailredirects:edit"
+    delete_url_name = "wagtailredirects:delete"
+    index_results_url_name = "wagtailredirects:index_results"
+    default_ordering = "old_path"
+    paginate_by = 20
+    is_searchable = True
+    page_title = gettext_lazy("Redirects")
+    search_fields = ["old_path", "redirect_page__url_path", "redirect_link"]
+    _show_breadcrumbs = True
+    columns = [
+        TitleColumn(
+            "old_path",
+            label=gettext_lazy("From"),
+            width="35%",
+            url_name="wagtailredirects:edit",
+            sort_key="old_path",
+        ),
+        Column(
+            "site",
+            label=gettext_lazy("Site"),
+            width="30%",
+        ),
+        RedirectPageColumn(
+            "redirect_page",
+            label=gettext_lazy("To"),
+            width="15%",
+        ),
+        StatusTagColumn(
+            "get_is_permanent_display",
+            label=gettext_lazy("Type"),
+            width="15%",
+        ),
+    ]
 
-    # Pagination
-    paginator = Paginator(redirects, per_page=20)
-    try:
-        redirects = paginator.page(request.GET.get("p", 1))
-    except InvalidPage:
-        raise Http404
+    @cached_property
+    def header_more_buttons(self):
+        buttons = super().header_more_buttons.copy()
 
-    # Render template
-    if request.headers.get("x-requested-with") == "XMLHttpRequest":
-        return TemplateResponse(
-            request,
-            "wagtailredirects/results.html",
-            {
-                "ordering": ordering,
-                "redirects": redirects,
-                "query_string": query_string,
-            },
+        buttons.clear()
+
+        buttons.append(
+            Button(
+                _("Import redirects"),
+                url=reverse("wagtailredirects:start_import"),
+                icon_name="doc-full-inverse",
+                priority=90,
+            )
         )
-    else:
-        return TemplateResponse(
-            request,
-            "wagtailredirects/index.html",
-            {
-                "ordering": ordering,
-                "redirects": redirects,
-                "query_string": query_string,
-                "search_form": SearchForm(
-                    data={"q": query_string} if query_string else None,
-                ),
-                "user_can_add": permission_policy.user_has_permission(
-                    request.user, "add"
-                ),
-            },
+        buttons.append(
+            Button(
+                _("Export redirects"),
+                url=reverse("wagtailredirects:report"),
+                icon_name="download",
+                priority=100,
+            )
         )
 
+        return buttons
+
```

Comment:
The optimisation was lost. Also, maybe we could use `select_related` instead of `prefetch_related`

```suggestion

def get_base_queryset(self):
    return super().get_base_queryset().select_related("redirect_page", "site")

```

Reply:
Sorry about that, it's fixed now. I don't know why I do these dumb things sometimes. 😅